prometheus/registry.go

```diff
@@ -448,12 +449,20 @@ func (r *Registry) Gather() ([]*dto.MetricFamily, error) {
 
 	collectWorker := func() {
 		for {
+			var metaMetricCollector Collector
 			select {
 			case collector := <-checkedCollectors:
-				collector.Collect(checkedMetricChan)
+				if _, ok := collector.(*MetaMetrics); ok {
+					metaMetricCollector = collector
+				} else {
+					collector.Collect(checkedMetricChan)
+				}
 			case collector := <-uncheckedCollectors:
 				collector.Collect(uncheckedMetricChan)
 			default:
+				if metaMetricCollector != nil {
+					metaMetricCollector.Collect(checkedMetricChan)
+				}
```

Comment:
oh, got it. we will come to this position only when we close checked and unchecked Metric Channels, but they can't be closed until `wg.Wait()` is called, `wg.Wait()` won't finish until `wg.Done()` is called, and so we get dead lock.

Reply:
it is not really fixed... it was called multiple time.
actual fix is around channel usage